### ISSUE TITLE
server.js sometimes crashes

### ISSUE THREAD

**Issue by jellybob** [cobyism/ghost-on-heroku#21]
Sorry this isn't more specific, I don't really use Node in production, and so I'm not sure how to go about debugging this. It happens on seemingly random occasions - sometimes almost as soon as the application is restarted, other times it'll run for days at a time.

```
2015-06-23T16:32:28.319398+00:00 heroku[web.1]: Starting process with command `npm start --production`
2015-06-23T16:32:31.173840+00:00 app[web.1]: Detected 512 MB available memory, 512 MB limit per process (WEB_MEMORY)
2015-06-23T16:32:31.173866+00:00 app[web.1]: Recommending WEB_CONCURRENCY=1
2015-06-23T16:32:38.321230+00:00 app[web.1]: 
2015-06-23T16:32:38.321234+00:00 app[web.1]: > ghost-on-heroku@0.6.2 start /app
2015-06-23T16:32:38.321236+00:00 app[web.1]: > node server.js
2015-06-23T16:32:38.321237+00:00 app[web.1]: 
2015-06-23T16:32:43.348030+00:00 app[web.1]: Migrations: Up to date at version 003
2015-06-23T16:32:44.103103+00:00 app[web.1]: Ghost is running... 
2015-06-23T16:32:44.103108+00:00 app[web.1]: Your blog is now available on http://52dishes.co.uk 
2015-06-23T16:32:44.103110+00:00 app[web.1]: Ctrl+C to shut down
2015-06-23T16:32:44.685499+00:00 heroku[web.1]: State changed from starting to up
2015-06-23T16:33:29.841782+00:00 app[web.1]: 
2015-06-23T16:33:29.842354+00:00 app[web.1]:         throw er; // Unhandled 'error' event
2015-06-23T16:33:29.842704+00:00 app[web.1]:               ^
2015-06-23T16:33:29.844019+00:00 app[web.1]: error: terminating connection due to administrator command
2015-06-23T16:33:29.844022+00:00 app[web.1]:     at Connection.parseE (/app/node_modules/ghost/node_modules/pg/lib/connection.js:534:11)
2015-06-23T16:33:29.844023+00:00 app[web.1]:     at Connection.parseMessage (/app/node_modules/ghost/node_modules/pg/lib/connection.js:361:17)
2015-06-23T16:33:29.844027+00:00 app[web.1]:     at Socket.emit (events.js:95:17)
2015-06-23T16:33:29.844029+00:00 app[web.1]:     at Socket.emit (events.js:92:17)
2015-06-23T16:33:29.842022+00:00 app[web.1]: events.js:72
2015-06-23T16:33:29.844028+00:00 app[web.1]:     at Socket.<anonymous> (_stream_readable.js:765:14)
2015-06-23T16:33:29.881234+00:00 app[web.1]: npm ERR! ghost-on-heroku@0.6.2 start: `node server.js`
2015-06-23T16:33:29.844032+00:00 app[web.1]:     at emitReadable (_stream_readable.js:423:5)
2015-06-23T16:33:29.844025+00:00 app[web.1]:     at Socket.<anonymous> (/app/node_modules/ghost/node_modules/pg/lib/connection.js:105:22)
2015-06-23T16:33:29.844034+00:00 app[web.1]:     at readableAddChunk (_stream_readable.js:166:9)
2015-06-23T16:33:29.844031+00:00 app[web.1]:     at emitReadable_ (_stream_readable.js:427:10)
2015-06-23T16:33:29.881705+00:00 app[web.1]: npm ERR! Failed at the ghost-on-heroku@0.6.2 start script.
2015-06-23T16:33:29.881444+00:00 app[web.1]: npm ERR! Exit status 8
2015-06-23T16:33:29.882201+00:00 app[web.1]: npm ERR! This is most likely a problem with the ghost-on-heroku package,
2015-06-23T16:33:29.882626+00:00 app[web.1]: npm ERR! You can get their info via:
2015-06-23T16:33:29.882523+00:00 app[web.1]: npm ERR!     node server.js
2015-06-23T16:33:29.883056+00:00 app[web.1]: npm ERR! System Linux 3.13.0-49-generic
2015-06-23T16:33:29.844035+00:00 app[web.1]:     at Socket.Readable.push (_stream_readable.js:128:10)
2015-06-23T16:33:29.872578+00:00 app[web.1]: 
2015-06-23T16:33:29.881599+00:00 app[web.1]: npm ERR! 
2015-06-23T16:33:29.882307+00:00 app[web.1]: npm ERR! not with npm itself.
2015-06-23T16:33:29.883780+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2015-06-23T16:33:29.882414+00:00 app[web.1]: npm ERR! Tell the author that this fails on your system:
2015-06-23T16:33:29.882767+00:00 app[web.1]: npm ERR!     npm owner ls ghost-on-heroku
2015-06-23T16:33:29.883202+00:00 app[web.1]: npm ERR! command "/app/.heroku/node/bin/node" "/app/.heroku/node/bin/npm" "start" "--production"
2015-06-23T16:33:29.882877+00:00 app[web.1]: npm ERR! There is likely additional logging output above.
2015-06-23T16:33:29.886787+00:00 app[web.1]: npm ERR! Additional logging details can be found in:
2015-06-23T16:33:29.886897+00:00 app[web.1]: npm ERR!     /app/npm-debug.log
2015-06-23T16:33:29.887282+00:00 app[web.1]: npm ERR! not ok code 0
2015-06-23T16:33:29.883340+00:00 app[web.1]: npm ERR! cwd /app
2015-06-23T16:33:29.883509+00:00 app[web.1]: npm ERR! node -v v0.10.38
2015-06-23T16:33:29.883647+00:00 app[web.1]: npm ERR! npm -v 1.4.28
2015-06-23T16:33:29.886671+00:00 app[web.1]: npm ERR! 
2015-06-23T16:33:30.678203+00:00 heroku[web.1]: Process exited with status 1
2015-06-23T16:33:30.691157+00:00 heroku[web.1]: State changed from up to crashed
```

**Comment by shea256** [comment]
Yes, I'm getting the same exact issue, actually. Trying to figure out what is going wrong.

**Comment by cobyism** [comment]
I’ve just pulled the logs on a site I’ve had running myself, and can confirm that I’m seeing the same issue. I’ll post an update if I can work out what’s going on.

**Comment by shea256** [comment]
@cobyism Thank you. Do you know if this is an issue that recently started coming up?

Here are my uptime / downtime logs for reference:

![ccufzhxqeqhoxp4yrddisqllcpk4cdv2xmezfskz4wk](https://cloud.githubusercontent.com/assets/994646/8362421/b2ab29f8-1b46-11e5-8905-fad25421bc3a.png)

**Comment by cobyism** [comment]
> Do you know if this is an issue that recently started coming up?

Unfortunately, no—I don’t have historical logs for any of the blogs I use this for. I found a few other similar reports (some as far back as 2011) in various places of the "terminating connection due to administrator command" happening randomly being an issue for people using Heroku in general:

- https://discussion.heroku.com/t/increase-in-pg-errors-fatal-terminating-connection-due-to-administrator-command/1027
- http://stackoverflow.com/questions/5787066/whats-the-cause-of-pgerror-fatal-terminating-connection-due-to-administrator
- https://github.com/dmclain/scrapy-heroku/issues/2
- https://twitter.com/hughevans/status/319576040429199361

… so I’m sceptical there’s anything specific to anything in this individual project.

Given the official [Heroku docs](https://devcenter.heroku.com/articles/postgres-logs-errors#fatal-terminating-connection-due-to-administrator-command) for the error in question, here’s some wild speculation:

- Some people at Heroku are performing lots of database maintenance.
- This could be something to do with the [upcoming new pricing](https://blog.heroku.com/archives/2015/5/7/new-dyno-types-public-beta), whereby hobby-tier dynos are forced to sleep a certain number of hours every day, thus the "administrator command" aspect of the error.
- :alien::alien::alien: 

I’m going to keep digging, because this is a crappy thing to have happen, however I’m going to close this issue because I don’t see anything that points to this being specifically connected to this project (nor upstream Ghost).

**Comment by jellybob** [comment]
I've migrated over to ghost.io because of this, but if its any help I can rule out the new pricing model as the cause for this, since the blog in question was running on the $7 hobby tier (hence moving to ghost.io - for an extra dollar a month I'm pretty happy to have file hosting and admin handled for me).

**Comment by cobyism** [comment]
I contacted Heroku support about this, and the upshot is that one of the limitations of the hobby tier postgres databases is unannounced maintenance, and also the potential for other apps ("noisy neighbour"s in Heroku’s words) to cause connection or load spikes (since many hobby database are run on a single shared server), which can cause disconnections like this.

TL;DR: Pay for a non-hobby database, or host Ghost somewhere else. :persevere:

**Comment by shea256** [comment]
OK wow, thank you. I upgraded and haven't noticed any issues.

**Comment by Fredric** [comment]
I have the same issue but i have something to add. 
After i started getting a lot of these errors on one app.( I get the error maybe 30 times a day at least), i simply created a new app with the exact same code base. This one works flawlessly. 100% uptime for several days now. 

Comparing at the apps postgres url's they also seem to be on the same server. (They are very similar)

It seems like this could be a case of good/bad luck but it sure feels weird. 

shea256: DId you upgrade your postgres to hobby?

**Comment by brightbits** [comment]
I think the main issue is, ghost crashes when the database connection dies, it should definitely handle that gracefully and reconnect?
